geoana/em/static/halfspace.py

```diff
@@ -0,0 +1,270 @@
+
+    @property
+    def rho(self):
+        """Resistivity in the point current in :math:`\\Omega \\cdot m`.
+
+        Returns
+        -------
+        float
+            Resistivity in the point current in :math:`\\Omega \\cdot m`.
+        """
+        return self._rho
+
+    @rho.setter
+    def rho(self, value):
+
+        try:
+            value = float(value)
+        except:
+            raise TypeError(f"current must be a number, got {type(value)}")
+
+        if value <= 0.0:
+            raise ValueError("current must be greater than 0")
+
+        self._rho = value
+
+    @property
+    def location(self):
+        """Location of observer in 3D space.
+
+        Returns
+        -------
+        (3) numpy.ndarray of float
+            Location of observer in 3D space. Default = np.r_[0,0,0].
+        """
+        return self._location
+
+    @location.setter
+    def location(self, vec):
+
+        try:
+            vec = np.atleast_1d(vec).astype(float)
+        except:
+            raise TypeError(f"location must be array_like, got {type(vec)}")
+
+        if len(vec) != 3:
+            raise ValueError(
+                f"location must be array_like with shape (3,), got {len(vec)}"
+            )
+
+        self._location = vec
+
+    def potential(self, xyz):
+        """Electric potential for a point current in a halfspace.
+
+        .. math::
+
+            V = \\frac{\\rho I}{2 \\pi R}
+
+        Parameters
+        ----------
+        xyz : (..., 3) numpy.ndarray
+            Locations to evaluate at in units m.
+
+        Returns
+        -------
+        V : (..., ) np.ndarray
+            Electric potential of point current in units V.
+
+        Examples
+        --------
+        Here, we define a point current with current=1A in a halfspace and plot the electric
+        potential as a function of distance.
+
+        >>> import numpy as np
+        >>> import matplotlib.pyplot as plt
+        >>> from geoana.em.static import PointCurrentWholeSpace
+
+        Define the point current.
+
+        >>> rho = 1.0
+        >>> current = 1.0
+        >>> simulation = PointCurrentWholeSpace(
+        >>>     current=current, rho=rho, location=None,
+        >>> )
+
+        Now we create a set of gridded locations, take the distances and compute the electric potential.
+
+        >>> X, Y = np.meshgrid(np.linspace(-1, 1, 20), np.linspace(-1, 1, 20))
+        >>> Z = np.zeros_like(X) + 0.25
+        >>> xyz = np.stack((X, Y, Z), axis=-1)
+        >>> r = np.linalg.norm(xyz, axis=-1)
+        >>> v = simulation.potential(xyz)
+
+        Finally, we plot the electric potential as a function of distance.
+
+        >>> plt.plot(r, v)
+        >>> plt.xlabel('Distance from point current')
+        >>> plt.ylabel('Electric potential')
+        >>> plt.title('Electric Potential as a function of distance from Point Current in a Halfspace')
+        >>> plt.show()
+        """
+
+        r_vec = xyz - self.location
```

Comment:
this point you should probably add in the `check_xyz_dim` function to test the input is correct shaped.
Also, for the half space, you need to throw a `ValueError` if any of the observation locations are below the surface. i.e.
`if np.any(xyz[..., -1] > 0)`